File: clients/ios/SitAware/Models/Drawing.swift
import Foundation

// MARK: - GeoJSON Types

/// Minimal GeoJSON FeatureCollection — stores drawing geometry and overlays.
struct GeoJSONFeatureCollection: Codable, Sendable {
    let type: String // "FeatureCollection"
    var features: [GeoJSONFeature]

    init(features: [GeoJSONFeature] = []) {
        self.type = "FeatureCollection"
        self.features = features
    }

    static var empty: GeoJSONFeatureCollection {
        GeoJSONFeatureCollection(features: [])
    }
}

struct GeoJSONFeature: Codable, Sendable {
    let type: String // "Feature"
    let geometry: GeoJSONGeometry
    var properties: [String: AnyCodable]?

    init(geometry: GeoJSONGeometry, properties: [String: AnyCodable]? = nil) {
        self.type = "Feature"
        self.geometry = geometry
        self.properties = properties
    }
}

/// GeoJSON geometry types used by SitAware (Point, LineString, Polygon).
enum GeoJSONGeometry: Codable, Sendable {
    case point([Double])                // [lng, lat]
    case lineString([[Double]])         // [[lng, lat], ...]
    case polygon([[[Double]]])          // [[[lng, lat], ...], ...]

    enum CodingKeys: String, CodingKey {
        case type
        case coordinates
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        let type = try container.decode(String.self, forKey: .type)

        switch type {
        case "Point":
            let coords = try container.decode([Double].self, forKey: .coordinates)
            self = .point(coords)
        case "LineString":
            let coords = try container.decode([[Double]].self, forKey: .coordinates)
            self = .lineString(coords)
        case "Polygon":
            let coords = try container.decode([[[Double]]].self, forKey: .coordinates)
            self = .polygon(coords)
        default:
            throw DecodingError.dataCorruptedError(
                forKey: .type, in: container,
                debugDescription: "Unknown GeoJSON geometry type: \(type)")
        }
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        switch self {
        case .point(let coords):
            try container.encode("Point", forKey: .type)
            try container.encode(coords, forKey: .coordinates)
        case .lineString(let coords):
            try container.encode("LineString", forKey: .type)
            try container.encode(coords, forKey: .coordinates)
        case .polygon(let coords):
            try container.encode("Polygon", forKey: .type)
            try container.encode(coords, forKey: .coordinates)
        }
    }
}

// MARK: - Drawing

struct DrawingResponse: Codable, Sendable, Identifiable {
    let id: String
    let ownerId: String
    let username: String
    let displayName: String
    let name: String
    let geojson: GeoJSONFeatureCollection
    let createdAt: String
    let updatedAt: String
}

// MARK: - Requests

struct CreateDrawingRequest: Codable, Sendable {
    let name: String
    let geojson: GeoJSONFeatureCollection
}

struct UpdateDrawingRequest: Codable, Sendable {
    var name: String?
    var geojson: GeoJSONFeatureCollection?
}

struct ShareDrawingRequest: Codable, Sendable {
    var groupId: String?
    var recipientId: String?
}

// MARK: - Share Info

struct DrawingShareInfo: Codable, Sendable, Identifiable {
    var id: String { messageId }
    let type: String // "group" or "user"
    let name: String
    let sharedAt: String
    let messageId: String

    enum CodingKeys: String, CodingKey {
        case type, name, sharedAt, messageId
        // `id` in the JSON maps to the share's own identifier (group_id or user_id),
        // but we use messageId as our Identifiable id to avoid conflict.
        case shareTargetId = "id"
    }

    let shareTargetId: String
}

// MARK: - WebSocket Drawing Update

/// Server -> Client: a drawing was updated by its owner.
typealias WSDrawingUpdated = DrawingResponse

File: .claude/worktrees/wonderful-lehmann/services/api/internal/model/drawing.go
package model

import (
	"encoding/json"
	"time"

	"github.com/google/uuid"
)

// Drawing represents a map drawing (annotation) in the system.
type Drawing struct {
	ID        uuid.UUID       `json:"-"`
	OwnerID   uuid.UUID       `json:"-"`
	Name      string          `json:"-"`
	GeoJSON   json.RawMessage `json:"-"`
	CreatedAt time.Time       `json:"-"`
	UpdatedAt time.Time       `json:"-"`
}

// DrawingResponse is the JSON representation returned by the API.
type DrawingResponse struct {
	ID          uuid.UUID       `json:"id"`
	OwnerID     uuid.UUID       `json:"owner_id"`
	Username    string          `json:"username"`
	DisplayName string          `json:"display_name"`
	Name        string          `json:"name"`
	GeoJSON     json.RawMessage `json:"geojson"`
	CreatedAt   time.Time       `json:"created_at"`
	UpdatedAt   time.Time       `json:"updated_at"`
}

// DrawingShareInfo describes a single share target for a drawing.
type DrawingShareInfo struct {
	Type      string    `json:"type"` // "group" or "user"
	ID        uuid.UUID `json:"id"`
	Name      string    `json:"name"`
	SharedAt  time.Time `json:"shared_at"`
	MessageID uuid.UUID `json:"message_id"`
}

// DrawingWithUser is a join result containing drawing + owner details.
type DrawingWithUser struct {
	Drawing
	Username    string
	DisplayName *string
}

// ToResponse converts a DrawingWithUser to its API response.
func (d *DrawingWithUser) ToResponse() DrawingResponse {
	displayName := ""
	if d.DisplayName != nil {
		displayName = *d.DisplayName
	}

	return DrawingResponse{
		ID:          d.ID,
		OwnerID:     d.OwnerID,
		Username:    d.Username,
		DisplayName: displayName,
		Name:        d.Name,
		GeoJSON:     d.GeoJSON,
		CreatedAt:   d.CreatedAt,
		UpdatedAt:   d.UpdatedAt,
	}
}
